dbos/_client.py

```diff
@@ -449,3 +461,77 @@ async def fork_workflow_async(
             application_version=application_version,
         )
         return WorkflowHandleClientAsyncPolling[Any](forked_workflow_id, self._sys_db)
+
+    def read_stream(self, workflow_id: str, key: str) -> Generator[Any, Any, None]:
+        """
+        Read values from a stream as a generator.
+        This function reads values from a stream identified by the workflow_id and key,
+        yielding each value in order until the stream is closed or the workflow terminates.
+
+        Args:
+            workflow_id: The ID of the workflow that wrote to the stream
+            key: The stream key to read from
+
+        Yields:
+            The values written to the stream in order
+        """
+        offset = 0
+        while True:
+            try:
+                value = self._sys_db.read_stream(workflow_id, key, offset)
+                if value == _dbos_stream_closed_sentinel:
+                    break
+                yield value
+                offset += 1
+            except ValueError:
+                # Poll the offset until a value arrives or the workflow terminates
+                status = get_workflow(self._sys_db, workflow_id)
+                if status is None:
+                    break
+                if (
+                    status.status == WorkflowStatusString.SUCCESS.value
+                    or status.status == WorkflowStatusString.ERROR.value
+                ):
+                    break
+                time.sleep(1.0)
+                continue
+
+    async def read_stream_async(
+        self, workflow_id: str, key: str
+    ) -> AsyncGenerator[Any, None]:
+        """
+        Read values from a stream as an async generator.
+        This function reads values from a stream identified by the workflow_id and key,
+        yielding each value in order until the stream is closed or the workflow terminates.
+
+        Args:
+            workflow_id: The ID of the workflow that wrote to the stream
+            key: The stream key to read from
+
+        Yields:
+            The values written to the stream in order
+        """
+        offset = 0
+        while True:
+            try:
+                value = await asyncio.to_thread(
+                    self._sys_db.read_stream, workflow_id, key, offset
+                )
+                if value == _dbos_stream_closed_sentinel:
+                    break
+                yield value
```

Comment:
I also think it'd be nice to yield the offset too... or provide a way to read from an offset.

Reply:
That can be added later as a separate API call, not sure if needed from the interface